File: src/main/scala/abc157/E.py
class SegmentTree:

    def __init__(self, n, fn=lambda x, y: min(x, y), unity=(1 << 31) - 1):
        self.n = 1
        while self.n < n:
            self.n *= 2

        self.arr = [unity] * (2 * self.n - 1)
        self.fn = fn
        self.unity = unity

    def update(self, i, x):
        k = self.n + i - 1
        self.arr[k] = x

        while k > 0:
            k = (k - 1) // 2
            self.arr[k] = self.fn(self.arr[2 * k + 1], self.arr[2 * k + 2])

    def query(self, l, r):
        return self._inner_query(l, r, 0, 0, self.n)

    def _inner_query(self, l, r, k, il, ir):
        if ir <= l or r <= il:
            return self.unity

        elif l <= il and ir <= r:
            return self.arr[k]

        vl = self._inner_query(l, r, k * 2 + 1, il, (il + ir) // 2)
        vr = self._inner_query(l, r, k * 2 + 2, (il + ir) // 2, ir)
        return self.fn(vl, vr)

    def __str__(self):
        return self.arr[0: 2 * self.n].__str__()

N = int(input())
S = input()
Q = int(input())

segs = [SegmentTree(N, lambda x, y: x + y, 0) for _ in range(26)]

L = len(S)

for i in range(L):
    x = ord(S[i]) - 97
    segs[x].update(i, 1)

for i in range(Q):
    m, a, b = input().split()

    if m == "1":
        a = int(a) - 1
        b = ord(b) - 97
        for i in range(26):
            segs[i].update(a, 0)
        segs[b].update(a, 1)
    else:
        a = int(a) - 1
        b = int(b)
        cnt = 0
        for i in range(26):
            cnt += 1 if segs[i].query(a, b) > 0 else 0
        print(cnt)

File: src/main/scala/abc157/E.java
package abc157;

import java.io.InputStream;
import java.io.OutputStream;
import java.io.PrintWriter;
import java.util.Arrays;
import java.util.Scanner;
import java.util.function.BiFunction;

public class E {
    long MOD = (long) 1e9 + 7;

    InputStream in;
    OutputStream out;

    public E() {
        this(System.in, System.out);
    }

    public E(InputStream in, OutputStream out) {
        this.in = in;
        this.out = out;
    }

    void run() {
        Scanner sc = new Scanner(this.in);
        int n = sc.nextInt();
        String s = sc.next();
        int q = sc.nextInt();

        SegmentTree[] sg = new SegmentTree[26];
        for (int i = 0; i < 26; i++) {
            sg[i] = new SegmentTree(n, Long::sum, 0);
        }

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
            int c = s.charAt(i) - 'a';
            a[i] = c;
            sg[c].update(i, 1);
        }

        PrintWriter out = new PrintWriter(this.out);

        for (int i = 0; i < q; i++) {
            int t = sc.nextInt();

            if (t == 1) {
                int iq = sc.nextInt() - 1;
                int c = sc.next().charAt(0) - 'a';
                if (a[iq] == c) continue;
                sg[c].update(iq, 1);
                sg[a[iq]].update(iq, 0);
                a[iq] = c;
            } else {
                int lq = sc.nextInt() - 1, rq = sc.nextInt();
                int cnt = 0;
                for (int j = 0; j < 26; j++) {
                    cnt += sg[j].query(lq, rq) > 0 ? 1 : 0;
                }
                out.println(cnt);
            }

        }

        out.flush();
    }

    public class SegmentTree {
        int n;
        long[] arr;
        BiFunction<Long, Long, Long> fn;
        long unity;

        SegmentTree(int n, BiFunction<Long, Long, Long> fn, long unity) {
            this.n = 1;
            while(this.n < n) {
                this.n *= 2;
            }
            arr = new long[2 * this.n - 1];
            this.fn = fn;
            this.unity = unity;
        }

        SegmentTree(int n) {
            this(n, Math::min, 1L << 60);
        }

        void update(int i, long x) {
            int k = this.n + i - 1;
            this.arr[k] = x;

            while (k > 0) {
                k = (k - 1) / 2;
                this.arr[k] = this.fn.apply(this.arr[2 * k + 1], this.arr[2 * k + 2]);
            }
        }

        long query(int l, int r) {
            return this.inneQuery(l, r, 0, 0, this.n);
        }

        private long inneQuery(int l, int r, int k, int il, int ir) {
            if (ir <= l || r <= il) {
                return this.unity;
            } else if(l <= il && ir <= r) {
                return this.arr[k];
            }

            long vl = this.inneQuery(l, r, k * 2 + 1, il, (il + ir) / 2);
            long vr = this.inneQuery(l, r, k * 2 + 2, (il + ir) / 2, ir);
            return this.fn.apply(vl, vr);
        }
    }

    void debug(Object... os) {
        System.err.println(Arrays.deepToString(os));
    }

    public static void main(String[] args) {
        new E().run();
    }
}